### PR TITLE
ENH: change 'times' attr from dictionary to list of tuples

### PR DESCRIPTION
**Pull request by st3107**
I change the "times" attribute in the FromEventStream and ToEventStream from dictionary to list. When "emit", the times are a tuple of time and start uid. When "stop", the times become a list of dictionary with key "time", "uid", "node".

Reason: In this way we can aviod using float as key in the stop doc so that this doc can be injected to db.

### CODE REVIEW

**Comment by sbillinge** [comment]
OK with me.  @CJ-Wright do you see any reason not to do this?

**Comment by st3107** [comment]
Add: the motivation to do this is that I test the example in the SHED tutorial and I found that the doc coming out of the FromEventStream cannot be injected into the database by using "db.insert". The cause is that we have float keys and when it is injected into mongodb, the mongo will convert it to string and then it will find that there is a "." in the string which is not allowed for the keys in mongodb. So we should avoid using float as keys.

The detail of the error report:

```
InvalidDocument                           Traceback (most recent call last)
<ipython-input-1-97e6c0ac8dd3> in <module>
     45 
     46 # Run the scan
---> 47 RE(bp.count([hw.ab_det], 5))

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/run_engine.py in __call__(self, *args, **metadata_kw)
    805             self._task_fut.add_done_callback(set_blocking_event)
    806 
--> 807         self._resume_task(init_func=_build_task)
    808 
    809         if self._interrupted:

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/run_engine.py in _resume_task(self, init_func)
    929                     if (exc is not None
    930                             and not isinstance(exc, _RunEnginePanic)):
--> 931                         raise exc
    932 
    933     def install_suspender(self, suspender):

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/run_engine.py in _run(self)
   1498             exit_reason = str(err)
   1499             self.log.exception("Run aborted")
-> 1500             raise err
   1501         finally:
   1502             if not exit_reason:

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/run_engine.py in _run(self)
   1363                     else:
   1364                         try:
-> 1365                             msg = self._plan_stack[-1].send(resp)
   1366                         # We have exhausted the top generator
   1367                         except StopIteration:

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/plans.py in count(detectors, num, delay, per_shot, md)
     75                                       num=num, delay=delay))
     76 
---> 77     return (yield from inner_count())
     78 
     79 

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/utils.py in dec_inner(*inner_args, **inner_kwargs)
   1044                 plan = gen_func(*inner_args, **inner_kwargs)
   1045                 plan = wrapper(plan, *args, **kwargs)
-> 1046                 return (yield from plan)
   1047             return dec_inner
   1048         return dec

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/preprocessors.py in stage_wrapper(plan, devices)
    950         return (yield from plan)
    951 
--> 952     return (yield from finalize_wrapper(inner(), unstage_devices()))
    953 
    954 

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/preprocessors.py in finalize_wrapper(plan, final_plan, pause_for_debug)
    507     cleanup = True
    508     try:
--> 509         ret = yield from plan
    510     except GeneratorExit:
    511         cleanup = False

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/preprocessors.py in inner()
    948     def inner():
    949         yield from stage_devices()
--> 950         return (yield from plan)
    951 
    952     return (yield from finalize_wrapper(inner(), unstage_devices()))

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/utils.py in dec_inner(*inner_args, **inner_kwargs)
   1044                 plan = gen_func(*inner_args, **inner_kwargs)
   1045                 plan = wrapper(plan, *args, **kwargs)
-> 1046                 return (yield from plan)
   1047             return dec_inner
   1048         return dec

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/preprocessors.py in run_wrapper(plan, md)
    325     yield from contingency_wrapper(plan,
    326                                    except_plan=except_plan,
--> 327                                    else_plan=close_run)
    328     return rs_uid
    329 

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/preprocessors.py in contingency_wrapper(plan, except_plan, else_plan, final_plan, pause_for_debug)
    587     else:
    588         if else_plan:
--> 589             yield from else_plan()
    590     finally:
    591         # if the exception raised in `GeneratorExit` that means

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/plan_stubs.py in close_run(exit_status, reason)
    843     :func:`bluesky.plans_stubs.open_run`
    844     """
--> 845     return (yield Msg('close_run', exit_status=exit_status, reason=reason))
    846 
    847 

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/run_engine.py in _run(self)
   1423                         # exceptions (coming in via throw) can be
   1424                         # raised
-> 1425                         new_response = await coro(msg)
   1426 
   1427                     # special case `CancelledError` and let the outer

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/run_engine.py in _close_run(self, msg)
   1630                                          "received before the 'open_run' "
   1631                                          "message") from ke
-> 1632         ret = (await current_run.close_run(msg))
   1633         del self._run_bundlers[run_key]
   1634         return ret

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/bundlers.py in close_run(self, msg)
    113             num_events=num_events,
    114         )
--> 115         await self.emit(DocumentNames.stop, doc)
    116         doc_logger.debug("[stop] document is emitted (run_uid=%r)", self._run_start_uid,
    117                          extra={'doc_name': 'stop',

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/run_engine.py in emit(self, name, doc)
   2270 
   2271     async def emit(self, name, doc):
-> 2272         self.emit_sync(name, doc)
   2273 
   2274 

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/run_engine.py in emit_sync(self, name, doc)
   2267         "Process blocking callbacks and schedule non-blocking callbacks."
   2268         schema_validators[name].validate(doc)
-> 2269         self.dispatcher.process(name, doc)
   2270 
   2271     async def emit(self, name, doc):

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/run_engine.py in process(self, name, doc)
   2290         doc : dict
   2291         """
-> 2292         exceptions = self.cb_registry.process(name, name.name, doc)
   2293         for exc, traceback in exceptions:
   2294             warn("A %r was raised during the processing of a %s "

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/utils.py in process(self, sig, *args, **kwargs)
    350             for cid, func in list(self.callbacks[sig].items()):
    351                 try:
--> 352                     func(*args, **kwargs)
    353                 except ReferenceError:
    354                     self._remove_proxy(func)

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/bluesky/utils.py in __call__(self, *args, **kwargs)
    431             mtd = self.func
    432         # invoke the callable and return the result
--> 433         return mtd(*args, **kwargs)
    434 
    435     def __eq__(self, other):

<ipython-input-1-97e6c0ac8dd3> in <lambda>(*x)
     42 RE.subscribe(db.insert)
     43 # have the RunEngine send data into the pipeline
---> 44 RE.subscribe(lambda *x: raw_source.emit(x))
     45 
     46 # Run the scan

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/rapidz/core.py in emit(self, x, asynchronous)
    395                 thread_state.asynchronous = True
    396             try:
--> 397                 result = self._emit(x)
    398                 if self.loop:
    399                     return gen.convert_yielded(result)

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/rapidz/core.py in _emit(self, x)
    376         result = []
    377         for downstream in list(self.downstreams):
--> 378             r = downstream.update(x, who=self)
    379             if type(r) is list:
    380                 result.extend(r)

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/shed/translation.py in update(self, x, who)
    127             self.times = {time.time(): doc["uid"]}
    128             self.start_uid = doc["uid"]
--> 129         return super().update(x, who=None)
    130 
    131 

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/shed/simple.py in update(self, x, who)
    672             # on their own time
    673             self.descriptor_uids = []
--> 674             [s.emit_stop(x) for s in self.subs]
    675         inner = doc.copy()
    676         if name == self.doc_type and (

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/shed/simple.py in <listcomp>(.0)
    672             # on their own time
    673             self.descriptor_uids = []
--> 674             [s.emit_stop(x) for s in self.subs]
    675         inner = doc.copy()
    676         if name == self.doc_type and (

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/shed/simple.py in emit_stop(self, x)
    257             self.incoming_start_uid = None
    258         stop = self.create_doc("stop", x)
--> 259         ret = self.emit(stop)
    260         [s.emit_stop(x) for s in self.subs]
    261         self.state = "stopped"

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/shed/translation.py in emit(self, x, asynchronous)
    213             self.times = {time.time(): self.start_uid}
    214         self.times[time.time()] = self.start_uid
--> 215         super().emit(x, asynchronous=asynchronous)
    216 
    217     def start_doc(self, x):

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/rapidz/core.py in emit(self, x, asynchronous)
    395                 thread_state.asynchronous = True
    396             try:
--> 397                 result = self._emit(x)
    398                 if self.loop:
    399                     return gen.convert_yielded(result)

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/rapidz/core.py in _emit(self, x)
    376         result = []
    377         for downstream in list(self.downstreams):
--> 378             r = downstream.update(x, who=self)
    379             if type(r) is list:
    380                 result.extend(r)

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/shed/translation.py in update(self, x, who)
    271                 )
    272             doc["graph"] = nx.node_link_data(graph)
--> 273         return self.emit((name, doc))
    274 
    275 

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/rapidz/core.py in emit(self, x, asynchronous)
    395                 thread_state.asynchronous = True
    396             try:
--> 397                 result = self._emit(x)
    398                 if self.loop:
    399                     return gen.convert_yielded(result)

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/rapidz/core.py in _emit(self, x)
    376         result = []
    377         for downstream in list(self.downstreams):
--> 378             r = downstream.update(x, who=self)
    379             if type(r) is list:
    380                 result.extend(r)

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/rapidz/core.py in update(self, x, who)
   1497     def update(self, x, who=None):
   1498         y = x + self.args
-> 1499         result = self.func(*y, **self.kwargs)
   1500         if gen.isawaitable(result):
   1501             return result

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/databroker/v1.py in insert(self, name, doc)
    962             "The method Broker.insert may be removed in a future release of "
    963             "databroker.", PendingDeprecationWarning)
--> 964         self._serializer(name, doc)
    965         # Make a reasonable effort to keep the Catalog in sync with new data.
    966         if name == 'stop':

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/suitcase/mongo_normalized/__init__.py in __call__(self, name, doc)
     90         # Python types compatible with pymongo.
     91         sanitized_doc = event_model.sanitize_doc(doc)
---> 92         return super().__call__(name, sanitized_doc)
     93 
     94     def update(self, name, doc):

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/event_model/__init__.py in __call__(self, name, doc, validate)
     74             instance as doc, a copy of doc, or a different dict altogether.
     75         """
---> 76         return self._dispatch(name, doc, validate)
     77 
     78     def _dispatch(self, name, doc, validate):

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/event_model/__init__.py in _dispatch(self, name, doc, validate)
     82         Optionally validate that the result is still a valid document.
     83         """
---> 84         output_doc = getattr(self, name)(doc)
     85 
     86         # If 'event' is not defined by the subclass but 'event_page' is, or

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/suitcase/mongo_normalized/__init__.py in stop(self, doc)
    168 
    169     def stop(self, doc):
--> 170         self._run_stop_collection.insert_one(doc)
    171 
    172     def __repr__(self):

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/pymongo/collection.py in insert_one(self, document, bypass_document_validation, session)
    696                          write_concern=write_concern,
    697                          bypass_doc_val=bypass_document_validation,
--> 698                          session=session),
    699             write_concern.acknowledged)
    700 

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/pymongo/collection.py in _insert(self, docs, ordered, check_keys, manipulate, write_concern, op_id, bypass_doc_val, session)
    610             return self._insert_one(
    611                 docs, ordered, check_keys, manipulate, write_concern, op_id,
--> 612                 bypass_doc_val, session)
    613 
    614         ids = []

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/pymongo/collection.py in _insert_one(self, doc, ordered, check_keys, manipulate, write_concern, op_id, bypass_doc_val, session)
    598 
    599         self.__database.client._retryable_write(
--> 600             acknowledged, _insert_command, session)
    601 
    602         if not isinstance(doc, RawBSONDocument):

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/pymongo/mongo_client.py in _retryable_write(self, retryable, func, session)
   1490         """Internal retryable write helper."""
   1491         with self._tmp_session(session) as s:
-> 1492             return self._retry_with_session(retryable, func, s, None)
   1493 
   1494     def _reset_server(self, address):

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/pymongo/mongo_client.py in _retry_with_session(self, retryable, func, session, bulk)
   1383                             raise last_error
   1384                         retryable = False
-> 1385                     return func(session, sock_info, retryable)
   1386             except ServerSelectionTimeoutError:
   1387                 if is_retrying():

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/pymongo/collection.py in _insert_command(session, sock_info, retryable_write)
    593                 session=session,
    594                 client=self.__database.client,
--> 595                 retryable_write=retryable_write)
    596 
    597             _check_write_command_response(result)

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/pymongo/pool.py in command(self, dbname, spec, slave_ok, read_preference, codec_options, check, allowable_errors, check_keys, read_concern, write_concern, parse_write_concern_error, collation, session, client, retryable_write, publish_events, user_fields)
    616         # Catch socket.error, KeyboardInterrupt, etc. and close ourselves.
    617         except BaseException as error:
--> 618             self._raise_connection_failure(error)
    619 
    620     def send_message(self, message, max_doc_size):

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/pymongo/pool.py in command(self, dbname, spec, slave_ok, read_preference, codec_options, check, allowable_errors, check_keys, read_concern, write_concern, parse_write_concern_error, collation, session, client, retryable_write, publish_events, user_fields)
    611                            use_op_msg=self.op_msg_enabled,
    612                            unacknowledged=unacknowledged,
--> 613                            user_fields=user_fields)
    614         except OperationFailure:
    615             raise

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/pymongo/network.py in command(sock, dbname, spec, slave_ok, is_mongos, read_preference, codec_options, session, client, check, allowable_errors, address, check_keys, listeners, max_bson_size, read_concern, parse_write_concern_error, collation, compression_ctx, use_op_msg, unacknowledged, user_fields)
    127         request_id, msg, size, max_doc_size = message._op_msg(
    128             flags, spec, dbname, read_preference, slave_ok, check_keys,
--> 129             codec_options, ctx=compression_ctx)
    130         # If this is an unacknowledged write then make sure the encoded doc(s)
    131         # are small enough, otherwise rely on the server to return an error.

~/anaconda3/envs/dev_pdfstream/lib/python3.7/site-packages/pymongo/message.py in _op_msg(flags, command, dbname, read_preference, slave_ok, check_keys, opts, ctx)
    702                 flags, command, identifier, docs, check_keys, opts, ctx)
    703         return _op_msg_uncompressed(
--> 704             flags, command, identifier, docs, check_keys, opts)
    705     finally:
    706         # Add the field back to the command.

InvalidDocument: key '1596648231.911391' must not contain '.'
```

**Comment by sbillinge** [comment]
@CJ-Wright do you see any problem with this?

**Comment by CJ-Wright** [comment]
Seems ok. You might consider adding a test that validates that SHED outputs are mongo friendly.

**Comment by st3107** [comment]
Good suggestion. Maybe I will add the example you used in the github page as a functionality test.

File: shed/tests/test_translation.py [hunk 1]
```diff
@@ -406,3 +407,17 @@ def test_dbfriendly(RE, hw):
     h2 = d[0].get("graph_hash")
     assert h1 != h2
     assert len(d) == 10 + 3
+
+
+def test_db_insertion(RE, hw):
+    db = Broker.named("temp")
```
Comment (CJ-Wright) [inline shed/tests/test_translation.py:413]: Does this fail on master branch?

File: .travis.yml [hunk 2]
```diff
@@ -34,6 +34,7 @@ before_install:
   - ./miniconda.sh -b -p ~/mc
   - export PATH=~/mc/bin:$PATH
   - conda config --set always_yes yes --set changeps1 no --set quiet true
+  - conda config --append channels nsls2forge
```
Comment (CJ-Wright) [inline .travis.yml:37]: Honestly someone should just fix the databroker recipe.

File: shed/tests/test_translation.py (hunk 1 above)
Comment (st3107) [reply shed/tests/test_translation.py:413]: I test this in the jupyter notebook. I failed because of the float number keys.

**Comment by st3107** [comment]
Weird. This "test_double_buffer_to_event_model_parallel" passed on my computer. However, it failed on CI because it took 5.04 s (> 0.5 * 10 s) to run.

File: .travis.yml (hunk 2 above)
Comment (st3107) [reply .travis.yml:37]: They moved all their packages to the nsls2forge channel.

**Comment by st3107** [comment]
Done. Ready to merge.

**Pull request merged**